File: product-service/src/main/java/ax/gritlab/buy_01/product/service/ProductService.java
package ax.gritlab.buy_01.product.service;

import ax.gritlab.buy_01.product.dto.ProductRequest;
import ax.gritlab.buy_01.product.dto.ProductResponse;
import ax.gritlab.buy_01.product.exception.ResourceNotFoundException;
import ax.gritlab.buy_01.product.exception.UnauthorizedException;
import ax.gritlab.buy_01.product.model.Product;
import ax.gritlab.buy_01.product.repository.ProductRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;

import java.time.ZonedDateTime;
import java.time.ZoneOffset;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.fasterxml.jackson.databind.node.ArrayNode;

@Service
@RequiredArgsConstructor
public class ProductService {
    // Delete all products for a user and publish product.deleted events
    public void deleteProductsByUserId(String userId) {
        List<Product> products = productRepository.findByUserId(userId);
        for (Product product : products) {
            List<String> mediaIds = product.getMediaIds();
            productRepository.delete(product);
            try {
                ObjectNode node = objectMapper.createObjectNode();
                node.put("id", product.getId());
                ArrayNode arr = node.putArray("mediaIds");
                if (mediaIds != null) {
                    for (String m : mediaIds) arr.add(m);
                }
                kafkaTemplate.send("product.deleted", objectMapper.writeValueAsString(node));
            } catch (Exception e) {
                kafkaTemplate.send("product.deleted", product.getId());
            }
        }
    }

    private final ProductRepository productRepository;
    private final RestTemplate restTemplate;
    private final org.springframework.kafka.core.KafkaTemplate<String, String> kafkaTemplate;
    private final ObjectMapper objectMapper;

    @Value("${media.service.url:http://media-service:8083/media}")
    private String mediaServiceUrl;

    @Value("${media.public.url:https://localhost:8443/api/media}")
    private String mediaPublicUrl;

    public List<ProductResponse> getAllProducts() {
        return productRepository.findAll().stream()
                .map(this::toProductResponse)
                .collect(Collectors.toList());
    }

    public ProductResponse getProductById(String id) {
        Product product = productRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Product not found with id: " + id));
        return toProductResponse(product);
    }

    public ProductResponse createProduct(ProductRequest request, String userId) {
        ZonedDateTime now = ZonedDateTime.now(ZoneOffset.UTC);
        Product product = Product.builder()
                .name(request.getName())
                .description(request.getDescription())
                .price(request.getPrice())
                .quantity(request.getQuantity())
                .userId(userId)
                .createdAt(now.toLocalDateTime())
                .updatedAt(now.toLocalDateTime())
                .build();
        Product saved = productRepository.save(product);
        return toProductResponse(saved);
    }

    public ProductResponse updateProduct(String id, ProductRequest request, String userId) {
        Product product = productRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Product not found with id: " + id));
        if (!product.getUserId().equals(userId)) {
            throw new UnauthorizedException("You do not have permission to update this product");
        }
        product.setName(request.getName());
        product.setDescription(request.getDescription());
        product.setPrice(request.getPrice());
        product.setQuantity(request.getQuantity());
        product.setUpdatedAt(ZonedDateTime.now(ZoneOffset.UTC).toLocalDateTime());
        Product saved = productRepository.save(product);
        return toProductResponse(saved);
    }

    public void deleteProduct(String id, String userId) {
        Product product = productRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Product not found with id: " + id));
        if (!product.getUserId().equals(userId)) {
            throw new UnauthorizedException("You do not have permission to delete this product");
        }
        List<String> mediaIds = product.getMediaIds();
        productRepository.delete(product);
        // Publish Kafka event for product deletion
        try {
            ObjectNode node = objectMapper.createObjectNode();
            node.put("id", id);
            ArrayNode arr = node.putArray("mediaIds");
            if (mediaIds != null) {
                for (String m : mediaIds) arr.add(m);
            }
            kafkaTemplate.send("product.deleted", objectMapper.writeValueAsString(node));
        } catch (Exception e) {
            kafkaTemplate.send("product.deleted", id);
        }
    }

    public ProductResponse associateMedia(String productId, String mediaId, String userId) {
        Product product = productRepository.findById(productId)
                .orElseThrow(() -> new ResourceNotFoundException("Product not found with id: " + productId));
        if (!product.getUserId().equals(userId)) {
            throw new UnauthorizedException("You do not have permission to modify this product");
        }
        product.getMediaIds().add(mediaId);
        Product saved = productRepository.save(product);

        // Call Media Service to update the productId in the media record
        try {
            String url = mediaServiceUrl + "/images/" + mediaId + "/product/" + productId + "?userId=" + userId;
            restTemplate.put(url, null);
        } catch (Exception e) {
            // Log the error but don't fail the product update
            System.err.println("Failed to update media productId: " + e.getMessage());
        }

        return toProductResponse(saved);
    }

    /**
     * Remove media ID from product's mediaIds array
     * Called by Media Service when media is deleted
     */
    public void removeMediaFromProduct(String productId, String mediaId) {
        Product product = productRepository.findById(productId)
                .orElseThrow(() -> new RuntimeException("Product not found"));

        product.getMediaIds().remove(mediaId);
        productRepository.save(product);
    }

    /**
     * Clean up all orphaned media IDs from products
     * This removes media IDs that no longer exist in the media database
     */
    public String cleanupOrphanedMedia() {
        List<Product> products = productRepository.findAll();
        int totalCleaned = 0;

        for (Product product : products) {
            List<String> validMediaIds = new ArrayList<>();

            // Check each media ID to see if it still exists
            for (String mediaId : product.getMediaIds()) {
                try {
                    // Try to call media service to check if media exists
                    String url = mediaServiceUrl + "/images/" + mediaId;
                    restTemplate.headForHeaders(url);
                    // If no exception, media exists
                    validMediaIds.add(mediaId);
                } catch (org.springframework.web.client.HttpClientErrorException e) {
                    // Media doesn't exist (404) or forbidden (403) - remove it
                    if (e.getStatusCode().value() == 404 || e.getStatusCode().value() == 403) {
                        System.out.println("Removing orphaned/inaccessible media ID: " + mediaId + " from product: "
                                + product.getId());
                        totalCleaned++;
                    } else {
                        // Other error - keep the media ID
                        validMediaIds.add(mediaId);
                    }
                } catch (Exception e) {
                    // Unknown error - keep the media ID to be safe
                    validMediaIds.add(mediaId);
                }
            }

            // Update product if any media IDs were removed
            if (validMediaIds.size() != product.getMediaIds().size()) {
                int removedCount = product.getMediaIds().size() - validMediaIds.size();
                product.setMediaIds(validMediaIds);
                productRepository.save(product);
                System.out.println(
                        "Cleaned product: " + product.getId() + " - Removed " + removedCount + " orphaned media IDs");
            }
        }

        return "Cleaned up " + totalCleaned + " orphaned media references from products";
    }

    /**
     * Convert Product entity to ProductResponse DTO with imageUrls
     */
    private ProductResponse toProductResponse(Product product) {
        // Convert mediaIds to image URLs using public URL for browser access
        List<String> imageUrls = product.getMediaIds().stream()
                .map(mediaId -> mediaPublicUrl + "/images/" + mediaId)
                .collect(Collectors.toList());

        return ProductResponse.builder()
                .id(product.getId())
                .name(product.getName())
                .description(product.getDescription())
                .price(product.getPrice())
                .stock(product.getQuantity())
                .sellerId(product.getUserId())
                .mediaIds(product.getMediaIds())
                .imageUrls(imageUrls)
                .createdAt(product.getCreatedAt() != null ? product.getCreatedAt().atZone(ZoneOffset.UTC).toString()
                        : null)
                .updatedAt(product.getUpdatedAt() != null ? product.getUpdatedAt().atZone(ZoneOffset.UTC).toString()
                        : null)
                .build();
    }
}

File: buy-01-ui/src/app/core/services/product.service.ts
import { Injectable, inject, signal } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, tap, map } from 'rxjs';
import { Auth } from './auth';
import { environment } from '../../../environments/environment';

export interface Product {
  id: string;
  name: string;
  description: string;
  price: number;
  stock?: number;
  sellerId?: string;
  mediaIds?: string[];
  imageUrls?: string[];
  createdAt?: string;
  updatedAt?: string;
}

// DTO for creating/updating products (matches backend ProductRequest)
export interface ProductRequest {
  name: string;
  description: string;
  price: number;
  quantity: number;
}

@Injectable({
  providedIn: 'root'
})
export class ProductService {
  private readonly http = inject(HttpClient);
  private readonly authService = inject(Auth);
  
  // API URL from environment configuration
  private readonly API_URL = environment.productsUrl;
  
  // Signals for state management
  private readonly productsSignal = signal<Product[]>([]);
  readonly products = this.productsSignal.asReadonly();
  
  /**
   * Get all products (public)
   * Calls backend API: GET /api/products
   */
  getAllProducts(): Observable<Product[]> {
    return this.http.get<Product[]>(this.API_URL).pipe(
      tap(products => this.productsSignal.set(products))
    );
  }
  
  /**
   * Get product by ID
   * Calls backend API: GET /api/products/{id}
   */
  getProductById(id: string): Observable<Product> {
    return this.http.get<Product>(`${this.API_URL}/${id}`);
  }
  
  /**
   * Get seller's products (authenticated)
   * CSR APPROACH: Get all products, filter client-side by sellerId
   */
  getSellerProducts(): Observable<Product[]> {
    const currentUserId = this.authService.currentUser()?.id;
    
    if (!currentUserId) {
      throw new Error('User not authenticated');
    }
    
    // Get all products and filter by sellerId (client-side filtering - CSR!)
    return this.http.get<Product[]>(this.API_URL).pipe(
      map(products => products.filter(p => p.sellerId === currentUserId))
    );
  }
  
  /**
   * Create product (sellers only)
   * Calls backend API: POST /api/products
   */
  createProduct(productRequest: ProductRequest): Observable<Product> {
    return this.http.post<Product>(this.API_URL, productRequest);
  }
  
  /**
   * Update product (sellers only - own products)
   * Calls backend API: PUT /api/products/{id}
   */
  updateProduct(id: string, productRequest: Partial<ProductRequest>): Observable<Product> {
    return this.http.put<Product>(`${this.API_URL}/${id}`, productRequest);
  }
  
  /**
   * Delete product (sellers only - own products)
   * Calls backend API: DELETE /api/products/{id}
   */
  deleteProduct(id: string): Observable<void> {
    return this.http.delete<void>(`${this.API_URL}/${id}`);
  }
  
  /**
   * Associate media with product
   * Calls backend API: POST /api/products/{productId}/media/{mediaId}
   */
  associateMedia(productId: string, mediaId: string): Observable<Product> {
    return this.http.post<Product>(`${this.API_URL}/${productId}/media/${mediaId}`, {});
  }




 /**
   * Remove media ID from product's mediaIds array
   * Calls backend API: DELETE /api/products/{productId}/remove-media/{mediaId}
   */
  removeMediaFromProduct(productId: string, mediaId: string): Observable<void> {
    return this.http.delete<void>(`${this.API_URL}/${productId}/remove-media/${mediaId}`);
  }









}
